### PR TITLE
[IOTDB-2569]Support ZSTD Compression

### PR DESCRIPTION
**Pull request by shuwenwei**
## Description
Add a new compression type Zstd.
```java
ZSTD(".zstd", (byte) 8);
```

### CODE REVIEW

**Comment by SteveYurongSu** [comment]
@shuwenwei Thanks for your great contribution!

I have two questions about the zstd.

1. How does the zstd algorithm compare with other algorithms?
2. I knew the zstd algorithm has a jni implementation and a non-jni implementation. Have we done any tests of the two implementations?

**Comment by shuwenwei** [comment]
> @shuwenwei Thanks for your great contribution!
> 
> I have two questions about the zstd.
> 
> 1. How does the zstd algorithm compare with other algorithms?
> 2. I knew the zstd algorithm has a jni implementation and a non-jni implementation. Have we done any tests of the two implementations?

image
image

**Comment by SteveYurongSu** [comment]
@shuwenwei great:)

@qiaojialin @xingtanzjr Interesting test results! It seems that GZIP (NOT SNAPPY) is a better choice for many scenarios.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/compress/IUnCompressor.java [hunk 1]
```diff
@@ -312,4 +315,45 @@ public CompressionType getCodecName() {
...
+  class ZstdUnCompressor implements IUnCompressor {
+
+    @Override
+    public int getUncompressedLength(byte[] array, int offset, int length) throws IOException {
+      return (int) Zstd.decompressedSize(array, offset, length);
+    }
+
+    @Override
+    public int getUncompressedLength(ByteBuffer buffer) throws IOException {
+      return (int) Zstd.decompressedSize(buffer);
```
Comment (SteveYurongSu) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/compress/IUnCompressor.java:328]: Is this a time-consuming call?

File: tsfile/src/main/java/org/apache/iotdb/tsfile/compress/IUnCompressor.java [hunk 2]
```diff
@@ -312,4 +315,45 @@ public CompressionType getCodecName() {
       return CompressionType.GZIP;
     }
   }
+
+  class ZstdUnCompressor implements IUnCompressor {
+
+    @Override
+    public int getUncompressedLength(byte[] array, int offset, int length) throws IOException {
+      return (int) Zstd.decompressedSize(array, offset, length);
```
Comment (SteveYurongSu) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/compress/IUnCompressor.java:323]: Is this a time-consuming call?

File: tsfile/src/main/java/org/apache/iotdb/tsfile/compress/ICompressor.java [hunk 3]
```diff
@@ -311,4 +316,61 @@ public CompressionType getType() {
...
+    public byte[] compress(byte[] data, int offset, int length) throws IOException {
+      if (data == null) {
+        return new byte[0];
+      }
+      byte[] toCompressData;
+      if (offset == 0 && length == data.length) {
+        toCompressData = data;
+      } else {
+        toCompressData = new byte[length];
+        System.arraycopy(data, offset, toCompressData, 0, length);
```
Comment (SteveYurongSu) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/compress/ICompressor.java:344]: I'm not sure. Is this necessary?

File: tsfile/src/main/java/org/apache/iotdb/tsfile/compress/IUnCompressor.java (hunk 2 above)
Comment (shuwenwei) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/compress/IUnCompressor.java:323]: It looks like the size is saved in the header.
image
image

File: tsfile/src/main/java/org/apache/iotdb/tsfile/compress/ICompressor.java (hunk 3 above)
Comment (shuwenwei) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/compress/ICompressor.java:344]: another implementation
image

**Review by SteveYurongSu — APPROVED** [review]
Overall LGTM. Thanks for your contribution :)

File: LICENSE-binary [hunk 4]
```diff
@@ -280,6 +280,11 @@ org.jline:jline:3.21.0
 com.github.luben:zstd-jni:1.5.4-2
 
 
+BSD 2-Clause
+------------
+com.github.luben:zstd-jni:1.5.4-2
```
Comment (SteveYurongSu) [inline LICENSE-binary:285]: BSD 2-Clause + BSD 3-Clause ?

Comment (shuwenwei) [reply LICENSE-binary]: BSD 2-Clause, I was wrong about that. The native Zstd is under 3-clause BSD licence. The JNI bindings is licenced under 2-clause BSD license.

**Review by SteveYurongSu — APPROVED** [review]

**Review by HTHou — APPROVED** [review]

**Review by xingtanzjr — APPROVED** [review]

**Pull request merged**